README.md
# TCard

Tinder like cards.

[![GitHub stars](https://img.shields.io/github/stars/xrr2016/tcard)](https://github.com/xrr2016/tcard/stargazers) [![pub package](https://img.shields.io/pub/v/tcard.svg)](https://pub.dev/packages/tcard) ![Test](https://github.com/xrr2016/tcard/workflows/Test/badge.svg)

- [TCard](#tcard)
  - [Install](#install)
  - [Usage](#usage)
    - [Normal widget](#normal-widget)
    - [Network image](#network-image)
    - [Use a controller to control](#use-a-controller-to-control)
    - [Determine the sliding direction](#determine-the-sliding-direction)
    - [Reset with new cards](#reset-with-new-cards)
  - [Property](#property)
  - [Contribute](#contribute)
  - [License](#license)

## Install

```yaml
# pubspec.yaml
dependencies:
  tcard: ^1.3.5
```

## Usage

### Normal widget

```dart
List<Widget> cards = List.generate(
  5,
  (index) => Container(
    color: Colors.blue,
    child: Center(
      child: Text(
        '$index',
        style: TextStyle(fontSize: 60, color: Colors.white),
      ),
    ),
  ),
);

TCard(
  cards: cards,
)
```

<img src="./example/demo.gif" width="334" alt="demo">

### Network image

```dart
List<String> images = [
  'https://gank.io/images/5ba77f3415b44f6c843af5e149443f94',
  'https://gank.io/images/02eb8ca3297f4931ab64b7ebd7b5b89c',
  'https://gank.io/images/31f92f7845f34f05bc10779a468c3c13',
  'https://gank.io/images/b0f73f9527694f44b523ff059d8a8841',
  'https://gank.io/images/1af9d69bc60242d7aa2e53125a4586ad',
];

List<Widget> cards = List.generate(
  images.length,
  (int index) {
    return Container(
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(16.0),
        boxShadow: [
          BoxShadow(
            offset: Offset(0, 17),
            blurRadius: 23.0,
            spreadRadius: -13.0,
            color: Colors.black54,
          )
        ],
      ),
      child: ClipRRect(
        borderRadius: BorderRadius.circular(16.0),
        child: Image.network(
          images[index],
          fit: BoxFit.cover,
        ),
      ),
    );
  },
);

TCard(
  size: Size(400, 600),
  cards: cards,
);
```

![images](./example/images.gif)

Image from [gank.io](gank.io)

### Use a controller to control

```dart
class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  TCardController _controller = TCardController();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            TCard(
              cards: cards,
              size: Size(360, 480),
              controller: _controller,
              onForward: (index, info) {
                print(index);
              },
              onBack: (index, info) {
                print(index);
              },
              onEnd: () {
                print('end');
              },
            ),
            SizedBox(
              height: 40,
            ),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: <Widget>[
                OutlineButton(
                  onPressed: () {
                    print(_controller);
                    _controller.back();
                  },
                  child: Text('Back'),
                ),
                OutlineButton(
                  onPressed: () {
                    _controller.reset();
                  },
                  child: Text('Reset'),
                ),
                OutlineButton(
                  onPressed: () {
                    _controller.forward();
                  },
                  child: Text('Forward'),
                ),
              ],
            ),
          ],
        ),
      ),
    );
  }
}
```

![controller](./example/controller.gif)

### Determine the sliding direction

```dart
 TCard(
  cards: cards,
  size: Size(360, 480),
  controller: _controller,
  onForward: (index, info) {
    print(index);
    print(info.direction);
    if (info.direction == SwipeDirection.Right) {
      print('like');
    } else {
      print('dislike');
    }
  },
  onBack: (index, info) {
    print(index);
  },
  onEnd: () {
    print('end');
  },
)
```

![like](./example/like.png)

### Reset with new cards

```dart
List<Widget> newCards = [];

TCardController _controller = TCardController();

_controller.reset(cards: newCards);
```

## Property

| property      |       type        | default |                                                      description                                                       | required |
| :------------ | :---------------: | :-----: | :--------------------------------------------------------------------------------------------------------------------: | :------: |
| cards         |  `List<Widget>`   | `null`  |                                                      Render cards                                                      |  `true`  |
| size          |      `Size`       | `null`  |                                                       Card size                                                        | `false`  |
| controller    | `TCardController` | `null`  |                                                    Card controller                                                     | `false`  |
| onForward     | `ForwardCallback` | `null`  |                                               Forward animation callback                                               | `false`  |
| onBack        |  `BackCallback`   | `null`  |                                                Back animation callback                                                 | `false`  |
| onEnd         |   `EndCallback`   | `null`  |                                                  Forward end callback                                                  | `false`  |
| lockYAxis     |      `bool`       | `false` |                                                  Lock Y Axis Gesture                                                   | `false`  |
| slideSpeed    |     `double`      |  `20`   |                 How quick should it be slided? less is slower. 10 is a bit slow. 20 is a quick enough.                 | `false`  |
| delaySlideFor |       `int`       |  `500`  | How long does it have to wait until the next slide is sliable? less is quicker. 100 is fast enough. 500 is a bit slow. | `false`  |

## Contribute

1. Fork it (https://github.com/xrr2016/tcard.git)
2. Create your feature branch (git checkout -b feature/foo)
3. Commit your changes (git commit -am 'Add some foo')
4. Push to the branch (git push origin feature/foo)
5. Create a new Pull Request

## License

[MIT](./LICENSE)


CHANGELOG.md
## [1.3.5] - 2021.03.30

1. v1.3.5

## [1.3.4] - 2021.03.30

1. v1.3.4

## [1.3.3] - 2021.02.23

1. Use credentialJson 🏗

## [1.3.2] - 2021.02.23

1. Use master branch publisher 🏗

## [1.3.1] - 2021.02.23

1. Fix onBack callback info index 🐛

## [1.3.0] - 2021.02.23

1. Add lockYAxis property ✨
2. Fix back direction 🐛
3. Reset cards with new widgets ✨
4. Update docs 📝

## [1.2.2] - 2020.07.27

1. Added direction attribute to controller forward method ✨

## [1.2.1] - 2020.07.27

1. Export SwipInfo class. 🐛

## [1.2.0] - 2020.07.27

1. Record the card sliding direction, use in onForward callback. ✨
2. Update example. 📝

## [1.1.0] - 2020.07.10

1. Fix onForward in the last card doesn't work. 🐛
2. Make the onEnd be called when the last element of the list is dismissed. ✨
3. Update readme. 📝

## [1.0.2] - 2020.06.19

- Update readme add dependencies. 🚀

## [1.0.1] - 2020.06.16

- Update import path, add plugin desction. 🎉

## [1.0.0] - 2020.06.16

- Version 1.0.0 release! 🚀

## [0.1.1] - 2020.06.15

- Add network images demo. 🎉

## [0.1.0] - 2020.06.15

- Version 0.1.0 release! 🚀


example/images.dart
// @dart=2.9

import 'package:flutter/material.dart';
import 'package:tcard/tcard.dart';

List<String> images = [
  'https://gank.io/images/5ba77f3415b44f6c843af5e149443f94',
  'https://gank.io/images/02eb8ca3297f4931ab64b7ebd7b5b89c',
  'https://gank.io/images/31f92f7845f34f05bc10779a468c3c13',
  'https://gank.io/images/b0f73f9527694f44b523ff059d8a8841',
  'https://gank.io/images/1af9d69bc60242d7aa2e53125a4586ad',
];

List<Widget> cards = List.generate(
  images.length,
  (int index) {
    return Container(
      decoration: BoxDecoration(
        color: Colors.white,
        borderRadius: BorderRadius.circular(16.0),
        boxShadow: [
          BoxShadow(
            offset: Offset(0, 17),
            blurRadius: 23.0,
            spreadRadius: -13.0,
            color: Colors.black54,
          )
        ],
      ),
      child: ClipRRect(
        borderRadius: BorderRadius.circular(16.0),
        child: Image.network(
          images[index],
          fit: BoxFit.cover,
        ),
      ),
    );
  },
);

class TCardPage extends StatefulWidget {
  @override
  _TCardPageState createState() => _TCardPageState();
}

class _TCardPageState extends State<TCardPage> {
  TCardController _controller = TCardController();

  int _index = 0;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: Column(
          children: <Widget>[
            SizedBox(height: 120),
            TCard(
              size: Size(400, 600),
              cards: cards,
              controller: _controller,
              onForward: (index, info) {
                _index = index;
                setState(() {});
              },
              onBack: (index, info) {
                _index = index;
                setState(() {});
              },
              onEnd: () {
                print('end');
              },
            ),
            SizedBox(height: 20),
            Row(
              mainAxisAlignment: MainAxisAlignment.spaceAround,
              children: <Widget>[
                OutlineButton(
                  onPressed: () {
                    _controller.back();
                  },
                  child: Text('Back'),
                ),
                OutlineButton(
                  onPressed: () {
                    _controller.forward();
                  },
                  child: Text('Forward'),
                ),
                OutlineButton(
                  onPressed: () {
                    _controller.reset();
                  },
                  child: Text('Reset'),
                ),
              ],
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {},
        child: Text(_index.toString()),
      ),
    );
  }
}
